utility/TemplateUtil.hpp

```diff
@@ -0,0 +1,157 @@
+/**
+ *   Copyright 2016, Quickstep Research Group, Computer Sciences Department,
+ *   University of Wisconsin—Madison.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ **/
+
+#ifndef QUICKSTEP_UTILITY_TEMPLATE_UTIL_HPP_
+#define QUICKSTEP_UTILITY_TEMPLATE_UTIL_HPP_
+
+#include <tuple>
+#include <utility>
+
+namespace quickstep {
+
+/** \addtogroup Utility
+ *  @{
+ */
+
+/**
+ * @brief Represents a compile-time sequence of integers.
+ *
+ * Seq is defined here for C++11 compatibility. For C++14 and above,
+ * std::integer_sequence can be used to achieve the same functionality.
+ */
+template<std::size_t ...>
+struct Seq {};
+
+/**
+ * @brief The helper class for creating Seq. GenSeq<N>::type is equivalent to
+ *        Seq<1,2,...,N>.
+ *
+ * GenSeq is defined here for C++11 compatibility. For C++14 and above,
+ * std::make_integer_sequence can be used to achieve the same functionality.
+ */
+template<std::size_t N, std::size_t ...S>
+struct GenSeq : GenSeq<N-1, N-1, S...> {};
```

Comment:
A better name `GeneratedSequence`?

FYI, `/CMakeLists.txt` detects whether a compiler supports `cxx14`. Therefore, you may want to add a `TODO` or directly use `std::make_integer_sequence` if having `cxx14` support. You could refer to `/threading/ThreadingConfig.h.in` for `QUICKSTEP_HAVE_CPP14_SHARED_TIMED_MUTEX`.